### ISSUE TITLE
Ideas

### ISSUE THREAD

**Issue by Silex** [MilestoneSystemsInc/PowerShellSamples#16]
Hello!

1. Add an ability to detect if we are already connected, right now `Get-ManagementServer` throw and it's impractical.
2. Add a `-Force` parameter to `Connect-ManagementServer`, right now it just warns but stays connected to the first server. Also I'd argue that it should be an error instead of a warning 😉

The idea for 1 & 2 is to simplify the following script:

``` ps1
# Bad, if DoStuff throws we never disconnect, then when we run again with another server the first one gets used!
Connect-ManagementServer -Server $server
DoStuff
Disconnect-ManagementServer

# Better
try {
  Connect-ManagementServer -Server $server
  DoStuff
}
finally {
  try { Disconnect-ManagementServer } catch {}
}

# The best solution IMHO, we don't really care if we don't disconnect as long as we are sure the connect does the right thing
Connect-ManagementServer -Server $server -Force
DoStuff
Disconnect-ManagementServer -Force
```

These kind of problems can bite you hard if you fool around in ISE and change `$server` and run and boom, you just modified the previously connected server.

And about my previous reports about `-ErrorAction` not being respected and `$ErrorActionPreference` working etc, it's all explained here https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1583 (with a more pragmatic approach explained here https://stackoverflow.com/questions/9294949/when-should-i-use-write-error-vs-throw-terminating-vs-non-terminating-errors/39949027#39949027)

Basically powershell is a real mess when it comes to errors.

**Comment by jhendricks123** [comment]
For the next release (possibly being pushed up to PSGallery tomorrow) I've made a number of changes including...

- Connect-ManagementServer now has a Force switch so Disconnect-ManagementServer will be called silently if needed before connecting to the provided management server
- Connect-ManagementServer without the Force switch will throw a terminating error if you're already connected to a management server (totally agree it's risky to just throw a warning there!)
- Disconnect-ManagementServer now returns silently if you're not already connected so you can safely call Disconnect whether you're connected or not.
- Get-ManagementServer writes an error instead of a terminating error if you're not connected to a management server. So you can more easily use Get-ManagementServer to test whether you're connected right now. Use it in a try/catch with ErrorAction Stop or test whether `Get-ManagementServer -ErrorAction Ignore` returns null
- An internal change in the way we keep track of the site we logged into makes it possible to use MilestonePSTools with runspaces to parallelize operations without having to use Connect-ManagementServer in each runspace/thread. Previously you might get an obnoxious "You're not connected" error even though it seems like you should have been. It doesn't help with PSJobs because those run in separate powershell.exe processes but runspaces running in the same process can now operate easily against the milestone connection established in the "parent thread".

Thanks for the feedback and apologies for the delay!

**Comment by jhendricks123** [comment]
Thanks again - closing this now with the release of MilestonePSTools 1.1.0 on PSGallery

**Comment by Silex** [comment]
Fantastic! Thanks.
